bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Display.java

```diff
@@ -613,6 +615,55 @@ public void asyncExec (Runnable runnable) {
 	}
 }
 
+/**
+ * Executes the given runnable in the user-interface thread of this Display.
+ * <ul>
+ * <li>If the calling thread is the user-interface thread of this display it is
+ * executed immediately and the method returns after the command has run, any
+ * runtime exception thrown will be rethrown at the caller</li>
+ * <li>In all other cases the the <code>run()</code> method of the runnable is
+ * asynchronously executed as with the method
+ * {@link Display#asyncExec(Runnable)} at the next reasonable opportunity. The
+ * caller of this method continues to run in parallel, and is not notified when
+ * the runnable has completed.</li>
+ * </ul>
+ * <p>
+ * This can be used in cases where one want to execute some piece of code that
+ * should be guaranteed to run in the user-interface thread regardless of the
+ * current thread.
+ * </p>
+ *
+ * <p>
+ * Note that at the time the runnable is invoked, widgets that have the receiver
+ * as their display may have been disposed. Therefore, it is advised to check
+ * for this case inside the runnable before accessing the widget.
+ * </p>
+ *
+ * @param runnable the runnable to execute in the user-interface thread, never
+ *                 <code>null</code>
+ * @throws RejectedExecutionException if this task cannot be accepted for
+ *                                    execution
+ * @throws NullPointerException       if runnable is null
+ */
+@Override
+public void execute(Runnable runnable) {
+	Objects.requireNonNull(runnable);
+	if (isDisposed()) {
+		throw new RejectedExecutionException(new SWTException (SWT.ERROR_WIDGET_DISPOSED, null));
+	}
+	if (thread == Thread.currentThread()) {
+		try {
+			runnable.run ();
```

Comment:
Judging from the logic in `syncExec` this appears to bypass a few callbacks, e.g. `Synchronizer.syncExec` will call `display.sendPreEvent(..)` and `sendPostEvent` respectively. 

Also this might cause reordering of events since the runnable will execute immediately on the DisplayThread ignoring other UI events that have been queued in the meantime. This doesn't feel right.

Reply:
> Also this might cause reordering of events since the runnable will execute immediately on the DisplayThread ignoring other UI events that have been queued in the meantime.

Never mind, this will happen with a syncExec issued from the DisplayThread anyhow. 

Instead of calling runnable.run I'd rather call syncExec if the current thread is the display thread.

Reply:
This is more to call the code of the runnable immediately without any extra work.

Reply:
You're saying you don't want to keep the contract of Display.syncExec if one uses the display as an Executor, but you do want to keep the contract for asyncExec?

Reply:
> You're saying you don't want to keep the contract of Display.syncExec if one uses the display as an Executor, but you do want to keep the contract for asyncExec?

The only contract is that the Executor guarantees is the given code is always executed in the UI thread. If the current Thread the UI Thread this contract is fulfilled by invoking the code right now. I don't see how `syncExec` would improve anything here?

Reply:
It would send the proper sendPre and sendPost event notification keeping all observers of these events up-to-date and not cloaking a random runnable as part of another runnable that's currently mid-flight. I don't understand why you would want to bypass these in favor of duplicating the exception handling here that is already done in the Synchronizer.

Reply:
> It would send the proper sendPre and sendPost event notification keeping all observers of these events up-to-date and not cloaking a random runnable as part of another runnable that's currently mid-flight. I

Why should this be desired/required? You seem to searching for problems that simply don't extis, if I have a method and call for example

```
label.setText("whatever")
label.setEnabled(true)
label.setWhatever()

```
no one would complain I should instead call `syncExec` for any single call...

Reply:
> no one would complain 

Neither would I.

I'd also not expect someone to write these three method calls as a sequence of futures based on the executor framework.

But if I see a sequence of futures that exercise the Display as an Executor, I'd expect the usual things to still work for the runnables that are submitted via that API, e.g. UI freeze monitoring. Can you guarantee that if you bypass the Synchronizer?

Reply:
As starting point, please look into org.eclipse.ui.internal.monitoring.EventLoopMonitorThreadManualTests.testFixedWork() and change `display.syncExec(doFixedAmountOfWork);` to something of the shape
```java
CompletableFuture<Void> future = CompletableFuture.runAsync(()->{}).thenApplyAsync(any->{
	return null;
}, display).thenRunAsync(doFixedAmountOfWork, display);
while(!future.isDone()) {
	while(display.readAndDispatch ()) {
		//dispatch
	}
}
```

You'll no longer see any stacks being reported from the long-running runnable.

Reply:
I can't think about a case where it makes a difference, maybe you have an example? For sure this dont prevent all possible deadlock and if one want can for sure deadlock as well as one can with the traditional API but that's also not the goal.

Reply:
> You'll no longer see any stacks being reported from the long-running runnable.

Long running runnable should never be executed in the UI thread so why should I want to do that? :-)

The desired use-case would be:

`CompletableFuture.supplyAsync(()->return doLenthyComputation()).thenAcceptAsync(result->updateTheUI(result), display);`

so no need for blocking programming style...

Reply:
I back off from this discussion. You're saying the pre and post events are useless, I'm saying the opposite. I don't see any progress. Since I'm not a committer on SWT, I let someone make the call.

Reply:
I'm not sure I understand the adamant resistance to what looks to me like a simple and constructive suggestion.   Is there some significant harm, or some violated principle, that comes from doing what's suggested?   These are after all methods being added to Display, so it would seem to me that the more these new methods behave (as expected) like the existing methods that they emulate or are intended to replace, the better.  

What exactly is the strong counter argument to not do what @szarnekow suggests?  I am a committer and I would do as he suggests.

Reply:
The general contract of the `Executor` interface:

>Executes the given command at some time in the future.  The command
> may execute in a new thread, in a pooled thread, or in the calling
> thread, at the discretion of the {@code Executor} implementation.

So we are actually free tom implement **what ever suffice**.

> I'm not sure I understand the adamant resistance to what looks to me like a simple and constructive suggestion.

Why is the current approach not "simple and constructive"? Why is it "adamant" if one likes to understand why we should choose a much more expensive construct that has a different semantics here?

> Is there some significant harm, or some violated principle, that comes from doing what's suggested?

Due to how SWT works, executing the code immediately or using syncExec has different semantics and different impact:

1. Using `syncExec` you might get code executed "in between" as per the API:  "At the time the runnable is invoked, widgets that have the receiver as their display may have been disposed", that might not be an issue, but could be an issue if such code is not just disposing a widget but changing its state (I think that what @szarnekow referes to with event reordering).
2. If you look at the implementation (e.g. GTK), there is a lot of overhead involved, that is calling the Synchronizer (what might or might not trigger more code, e.g Platform has `UISyncronizer` doing additional checks)), synchronizing on locks and performing system calls.
3. Whenever I see code in our platform code or other libs, the usual idiom is to check if one is on the UI thread if yes, execute the code, if no either `asyncExec` or `syncExec` code (depending on weather would want to be blocked or not), I never found an example where `syncExec` is used just randomly if one is already in the UI thread.

That's why I have given the example with just calling on (or more) methods that strictly requires a UI-Thread, something that is common, no one ever would argue you should use `syncExec` all the time to trigger events, nor one would suggest that it is Ok to run lengthy operation when using `syncExec` ...

So for me it sounds not that bad simply replicate this common pattern. If the other approach has strong benefits I'm open to change that, but it seems a very costly operation and I can't tell if the event reordering can cause any harm.

Whatever is decided here the implementation will still never:

1. Prevent the user from any deadlocks, as due the single-threaded nature you can always created a situation where deadlocks occur when you use blocking calls, **but** will give the user the opportunity to use non blocking programming paradigms.
2. Make it "save" to do heavy weight computations or long running task in the UI thread, **but**  will give the user the opportunity to simply fire this work "in the background" and using the result to do the necessary UI work as short as possible.

Reply:
Please ensure that UI freeze monitoring works for tasks submitted via the newly introduced method. I’m not asking for more than that. You’ll get it for free when using syncExec after checking that the current thread is the display thread.